RecoLocalCalo/HcalRecProducers/src/HBHERecHitProducerGPU.cc

```diff
@@ -0,0 +1,292 @@
+
+    configParameters_.tzeroTimeSlew = tzeroValues[HcalTimeSlew::Medium];
+    configParameters_.slopeTimeSlew = slopeValues[HcalTimeSlew::Medium];
+    configParameters_.tmaxTimeSlew = tmaxValues[HcalTimeSlew::Medium];
+
+    outputGPU_.allocate(configParameters_);
+    scratchGPU_.allocate(configParameters_);
+
+    // FIXME: use default device and default stream
+    cudaCheck( cudaMalloc((void**)&configParameters_.pulseOffsetsDevice,
+        sizeof(int) * configParameters_.pulseOffsets.size()) );
+    cudaCheck( cudaMemcpy(configParameters_.pulseOffsetsDevice,
+                          configParameters_.pulseOffsets.data(),
+                          configParameters_.pulseOffsets.size() * sizeof(int),
+                          cudaMemcpyHostToDevice) );
+}
+
+HBHERecHitProducerGPU::~HBHERecHitProducerGPU() {
+    outputGPU_.deallocate(configParameters_);
+    scratchGPU_.deallocate(configParameters_);
+
+    cudaCheck( cudaFree(configParameters_.pulseOffsetsDevice) );
+}
+
+void HBHERecHitProducerGPU::fillDescriptions(edm::ConfigurationDescriptions& cdesc) {
+    edm::ParameterSetDescription desc;
+    desc.add<uint32_t>("maxChannels", 10000u);
+    desc.add<uint32_t>("maxTimeSamples", 10);
+    desc.add<std::vector<int>>("pulseOffsets", {-3, -2, -1, 0, 1, 2, 3, 4});
+    desc.add<uint32_t>("kprep1dChannelsPerBlock", 32);
+    desc.add<edm::InputTag>("digisLabelF01HE", 
+        edm::InputTag{"hcalRawToDigiGPU", "f01HEDigisGPU"});
+    desc.add<edm::InputTag>("digisLabelF5HB", 
+        edm::InputTag{"hcalRawToDigiGPU", "f5HBDigisGPU"});
+    desc.add<std::string>("recHitsLabelM0HBHE", "recHitsM0HBHE");
+    desc.add<int>("sipmQTSShift", 0);
+    desc.add<int>("sipmQNTStoSum", 3);
+    desc.add<int>("firstSampleShift", 0);
+    desc.add<bool>("useEffectivePedestals", true);
+
+    desc.add<double>("meanTime", 0.f);
+    desc.add<double>("timeSigmaSiPM", 2.5f);
+    desc.add<double>("timeSigmaHPD", 5.0f);
+    desc.add<double>("ts4Thresh", 0.0);
+
+    desc.add<bool>("applyTimeSlew", true);
+    desc.add<std::vector<double>>("tzeroTimeSlewParameters", 
+        {23.960177, 11.977461, 9.109694});
+    desc.add<std::vector<double>>("slopeTimeSlewParameters", 
+        {-3.178648, -1.5610227, -1.075824});
+    desc.add<std::vector<double>>("tmaxTimeSlewParameters", 
+        {16.00, 10.00, 6.25});
+
+    std::string label = "hbheRecHitProducerGPU";
+    cdesc.add(label, desc);
+}
+
+void HBHERecHitProducerGPU::acquire(
+        edm::Event const& event,
+        edm::EventSetup const& setup,
+        edm::WaitingTaskWithArenaHolder holder) 
+{
+#ifdef HCAL_MAHI_CPUDEBUG
+    auto start = std::chrono::high_resolution_clock::now();
+#endif
+
+    // input + raii
+    auto const& f01HEProduct = event.get(digisTokenF01HE_);
+    auto const& f5HBProduct = event.get(digisTokenF5HB_);
+    CUDAScopedContextAcquire ctx{f01HEProduct, std::move(holder), cudaState_};
+    auto const& f01HEDigis = ctx.get(f01HEProduct);
```

Comment:
decoding is done according to the specs of the uhtr doc. it's not about hb or he here, it's about flavors. again, everythign is done according to the doc.

Reply:
and the cpu version, i preferred in the naming of collections not to use qie11/10 as it is not part of the uhtr specs

Reply:
The decoding from the HW should be done at digi step not done again at RECO, otherwise we do million times everything from scratch
Multiple instances of the same things is not maintainable.

Reply:
this is not what i meant. check HcalRawToDigiGPU producer in EventFilter/HcalRawToDigi which runs the decoding... it is run just once...

Reply:
Not much convinced since the indexing is done by geometry of the hb/he.
That is different between the Run2 and Run3.
So now we nest digi flavour and geometry indexing. in the CPU version I do not need to know nothing of those (imagine we will have another HEP17 equivalent pilot run).